File: javascript/07p.js
function transpose(mat) {
    var clone = [];
    mat.forEach(function(row) {
        clone.push(row.slice());
    });

    mat.length = clone[0].length;
    for(var y = 0; y < clone.length; y++) {
        for(var x = 0; x < clone[0].length; x++) {
            mat[x][y] = clone[y][x];
        }
    }
}

function solve(input) {
    var mat = [
        "414213".split(""),
        "732050".split(""),
        "236067".split(""),
        "645751".split(""),
        "316624".split(""),
        "605551".split(""),
    ];

    input.split("").forEach(function(s) {
        var i = "ABCDEFuvwxyz".indexOf(s);
        if(i < 6) {
            transpose(mat);
            mat.sort(function(a, b) { return a[i] - b[i] });
            transpose(mat);
        } else {
            mat.sort(function(a, b) { return a[i - 6] - b[i - 6] });
        }
    });

    return mat[0].join("");
}

function test(input, expected) {
    var result = solve(input);
    if(result !== expected) {
        console.log("-------------\ninput = %s\n\nexpected = %s\nresult   = %s", input, expected, result);
    }
}

test( "AvEx"     , "305027");
test( "A"        , "112344");
test( "C"        , "241413");
test( "F"        , "134214");
test( "u"        , "236067");
test( "w"        , "732050");
test( "y"        , "414213");
test( "yx"       , "732050");
test( "ux"       , "236067");
test( "EF"       , "131424");
test( "DF"       , "134124");
test( "Au"       , "055165");
test( "uA"       , "023667");
test( "By"       , "234114");
test( "yB"       , "114342");
test( "yBy"      , "357020");
test( "yByB"     , "350072");
test( "AuBvCw"   , "131244");
test( "FAuFBvFCw", "300527");
test( "AuBv"     , "112344");
test( "CwDx"     , "515056");
test( "FzyE"     , "324114");
test( "uAwDyB"   , "114324");
test( "zExCvF"   , "073520");
test( "uFxEv"    , "002357");
test( "DyCwB"    , "076362");

File: haskell/07p.hs
-- オフラインリアルタイムどう書く第7回の参考問題
-- XY-Sort
-- http://nabetani.sakura.ne.jp/hena/ord7xysort/

-- Haskellに標準で用意されているソート関数とtranspose(転置)関数を使って解きます。


import Data.List
import Data.Char

initMatrix :: [String]
initMatrix = [
    "414213",
    "732050",
    "236067",
    "645751",
    "316624",
    "605551"
    ]

sortY, sortX :: Int -> [String] -> [String]
sortY n mat = sortBy (\x y -> compare (x !! n) (y !! n)) mat
sortX n mat = transpose . (sortY n) . transpose $ mat

sortXY :: String -> [String] -> [String]
sortXY [] mat = mat
sortXY (x:xs) mat
    | isUpper x = sortXY xs $ sortX (ord x - ord 'A') mat
    | otherwise = sortXY xs $ sortY (ord x - ord 'u') mat

solve :: String -> String
solve input = head $ sortXY input initMatrix


test :: String -> String -> IO ()
test input expected
    | result == expected = return ()
    | otherwise          = error $ show input ++ " -- expected " ++ show expected ++ " but " ++ show result
    where
        result = solve input

main :: IO ()
main = do
    test "AvEx"      "305027"
    test "A"         "112344"
    test "C"         "241413"
    test "F"         "134214"
    test "u"         "236067"
    test "w"         "732050"
    test "y"         "414213"
    test "yx"        "732050"
    test "ux"        "236067"
    test "EF"        "131424"
    test "DF"        "134124"
    test "Au"        "055165"
    test "uA"        "023667"
    test "By"        "234114"
    test "yB"        "114342"
    test "yBy"       "357020"
    test "yByB"      "350072"
    test "AuBvCw"    "131244"
    test "FAuFBvFCw" "300527"
    test "AuBv"      "112344"
    test "CwDx"      "515056"
    test "FzyE"      "324114"
    test "uAwDyB"    "114324"
    test "zExCvF"    "073520"
    test "uFxEv"     "002357"
    test "DyCwB"     "076362"
